u2f-ref-code/java/src/com/google/u2f/tools/X509Util.java

```diff
@@ -0,0 +1,155 @@
+   * @return a parsed certificate chain
+   */
+  public static X509Certificate[] parseCertificateChain(byte[] encodedDerCertificates) {
+    return parseCertificateChain(
+        new DataInputStream(new ByteArrayInputStream(encodedDerCertificates)));
+  }
+
+  /**
+   * Parse a certificate chain out of a DER-encoded DataInputStream.  Will attempt to parse certs
+   * out of the stream until it can no longer do so.  Will return all the certs (in order) that were
+   * parsed successfully.
+   *
+   * @param encodedDerCertificates DER encoded X509 certificates
+   * @return a parsed certificate chain
+   */
+  public static X509Certificate[] parseCertificateChain(DataInputStream encodedDerCertificates) {
+    List<X509Certificate> certChain = new LinkedList<X509Certificate>();
+    try {
+      while (encodedDerCertificates.available() > 0) {
+        // Note: CertificateFactory#generateCertificate will extract one certificate from a byte
+        // stream even when there are trailing bytes remaining. Relying on this allows us to
+        // repeatedly call this method over a byte stream containing concatenated certificates.
+        certChain.add((X509Certificate) CertificateFactory.getInstance(X509).generateCertificate(
+            encodedDerCertificates));
+        Log.info("Parsed cert: " + certChain.get(certChain.size() - 1));
+      }
+    } catch (IOException | CertificateException e) {
+      // no more certs to read
+      Log.info("No more certs");
+    }
+    return certChain.toArray(new X509Certificate[0]);
+  }
+
+  /**
+   * Attempts to verify a certificate chain and makes sure that it chains up to at least one of the
+   * provided CA root certificates.
+   *
+   * @param certChain The certificate chain to validate.  The leaf certificate is assumed to be in
+   *   {@code certChain[0]} and it is assumed that {@code certChain[i]} certificate is signed by
+   *   {@code certChain[i+1]}.  Finally, {@code certChain[certChain.length-1]} should be signed
+   *   by one of the root CA certificates.
+   * @param caCerts The trusted root certificates.
+   * @return {@code true} if validation succeeded
+   * @throws U2FException if there was a parsing error
+   */
+  public static boolean verifyCertChain(X509Certificate[] certChain, X509Certificate[] caCerts)
+      throws U2FException {
+    if (caCerts == null || certChain == null || certChain.length == 0) {
+      return false;
+    }
+
+    // Walk through intermediates up the chain
+    try {
+      for (int i = 0; i < certChain.length - 1; i++) {
+        certChain[i].verify(certChain[i + 1].getPublicKey());
+      }
+    } catch (SignatureException | NoSuchProviderException | InvalidKeyException
+        | NoSuchAlgorithmException | CertificateException e) {
+      Log.log(Level.SEVERE, "Cannot validate cert chain is correctly signed by intermediaries", e);
+      return false;
+    }
+
+    // Now attempt to verify up to one of the roots
+    boolean validated = true;
```

Comment:
This default value, combined with the for loop that will not execute for an empty caCerts list, will allow the caller to get validated = true when the caCerts list is empty. Perhaps that's what you mean, but it certainly looks odd.